File: integration-tests-java/src/main/java/io/micronaut/jimmer/java/it/entity/BookStore.java
package io.micronaut.jimmer.java.it.entity;

import io.micronaut.jimmer.java.it.resolver.BookStoreAvgPriceResolver;
import java.math.BigDecimal;
import java.util.List;
import org.babyfish.jimmer.sql.*;
import org.jetbrains.annotations.Nullable;

@Entity
public interface BookStore extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key
    String name();

    @Nullable
    String website();

    @OneToMany(
            mappedBy = "store",
            orderedProps = {@OrderedProp("name"), @OrderedProp(value = "edition", desc = true)})
    List<Book> books();

    @Transient(value = BookStoreAvgPriceResolver.class)
    BigDecimal avgPrice();

    @Transient(ref = "bookStoreNewestBooksResolver")
    List<Book> newestBooks();
}

File: integration-tests-kotlin/src/main/kotlin/io/micronaut/jimmer/kotlin/it/entity/BookStore.kt
package io.micronaut.jimmer.kotlin.it.entity

import io.micronaut.jimmer.kotlin.it.resolver.BookStoreNewestBooksResolver
import org.babyfish.jimmer.sql.Entity
import org.babyfish.jimmer.sql.GeneratedValue
import org.babyfish.jimmer.sql.GenerationType
import org.babyfish.jimmer.sql.Id
import org.babyfish.jimmer.sql.OneToMany
import org.babyfish.jimmer.sql.OrderedProp
import java.math.BigDecimal

@Entity
interface BookStore : BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long

    val name: String

    val website: String?

    @OneToMany(mappedBy = "store", orderedProps = [OrderedProp("name"), OrderedProp(value = "edition", desc = true)])
    val books: List<Book>

    @org.babyfish.jimmer.sql.Transient(ref = "bookStoreAvgPriceResolver")
    val avgPrice: BigDecimal

    @org.babyfish.jimmer.sql.Transient(BookStoreNewestBooksResolver::class)
    val newestBooks: List<Book>
}
